### PR TITLE
expand implicit conversions for pointers to arrays

### PR DESCRIPTION
**Pull request by dtarditi**
This addresses issue #20 by expanding the implicit conversions for pointers to arrays.  We now allow conversions from pointers to unchecked arrays to pointers to checked arrays.   This is needed to support passing unchecked multi-dimensional arrays to functions that have checked array parameters.

This change adds a notion of assignment compatibility.  Compatibility of types in C is similar to equality and is not directional.   We need a directional conversion, where source and destination matter.  Informally, we allow a type A to be converted to another type B if A and B are the same except that B has more checking.  

I considered defining a recursive notion of assignment compatibility for referent types that are pointers, but decided not to do that now.    It requires other changes to the specification that go beyond what is needed for issue #20, so I'm putting that work off.

I also reworked some of the examples for casts and conversions and added non-examples of code that will fail to check.

### CODE REVIEW

File: spec/bounds_safety/interoperation.tex [hunk 1]
```diff
@@ -515,27 +586,29 @@ \section{Bounds-safe interfaces to existing unchecked functions and variables}
...
+      typehas an unchecked pointer type, the corresponding argument expression
+      has a checked pointer type, and the argument referent type is assignment
+      compatible with the parameter referent type, than the argument expression
+      will be converted implicitly to the unchecked pointer type.
 \item Assignments to a variable with external scope: if the variable being
-     assigned to has an unchecked pointer type and a bound-safe interface and the
-     right-hand side expression has a checked pointer whose referent type is
-     compatible with the
-     variable referent type, the right-hand side expression will be converted
+     assigned to has an unchecked pointer type and a bound-safe interface, the
```
Comment (awruef) [inline spec/bounds_safety/interoperation.tex]: bounds-safe interface?

File: spec/bounds_safety/interoperation.tex [hunk 2]
```diff
@@ -515,27 +586,29 @@ \section{Bounds-safe interfaces to existing unchecked functions and variables}
...
 \begin{itemize}
-\item Function call arguments: If the function being called has a 
-      bounds-safe interface for unchecked pointer type arguments, a parameter type
-      has an unchecked pointer type, and the corresponding argument expression has a checked
-      pointer type whose referent type is compatible with the parameter
-      referent type, the argument expression will be converted implicitly to the 
-      unchecked pointer type.
+\item Function call arguments: If the function being called has a
+      bounds-safe interface for unchecked pointer type arguments, a parameter
+      typehas an unchecked pointer type, the corresponding argument expression
```
Comment (awruef) [inline spec/bounds_safety/interoperation.tex]: type has?

File: spec/bounds_safety/interoperation.tex [hunk 3]
```diff
@@ -515,27 +586,29 @@ \section{Bounds-safe interfaces to existing unchecked functions and variables}
...
+      has a checked pointer type, and the argument referent type is assignment
+      compatible with the parameter referent type, than the argument expression
```
Comment (awruef) [inline spec/bounds_safety/interoperation.tex]: then the?

File: spec/bounds_safety/interoperation.tex [hunk 4]
```diff
@@ -515,27 +586,29 @@ \section{Bounds-safe interfaces to existing unchecked functions and variables}
...
+      will be converted implicitly to the unchecked pointer type.
 \item Assignments to a variable with external scope: if the variable being
-     assigned to has an unchecked pointer type and a bound-safe interface and the
-     right-hand side expression has a checked pointer whose referent type is
-     compatible with the
-     variable referent type, the right-hand side expression will be converted
+     assigned to has an unchecked pointer type and a bound-safe interface, the
+     right-hand side expression has a checked pointer type, and the right-hand
+     side expression referent type is assignment compatible with the referent
+     type of the variable, than the right-hand side expression will be converted
```
Comment (awruef) [inline spec/bounds_safety/interoperation.tex]: then the?

**Comment by awruef** [comment]
This makes sense to me. I found some very minor word-choice things and left comments for them.

**Comment by dtarditi** [comment]
I've fixed the word-choice issues that you pointed out.  Thanks!

**Pull request merged**